### ISSUE TITLE
iframeResizer, feedback popover, and mqeditor

### ISSUE THREAD

**Issue by Alex-Jordan** [openwebwork/pg#1041]
The new feedback mechanism is working very well with PTX. I removed all of the code from PTX's `pretext-webwork.js` that had to do with feedback, and can pretty much just rely on the webwork/pg feedback now. So that is pretty great.

One thing that is not working well yet is that when either a feedback popover or the mathquill menu is opened, the iframeResizer is not responding to this.

<img width="659" alt="Screenshot 2024-03-22 at 2 11 30 PM" src="https://github.com/openwebwork/webwork2/assets/4732672/9bc15dae-3a2e-43df-916f-f516aae3d02f">

iframeResizer is working for other things though, like you can see here when I expanded a solution:

<img width="653" alt="Screenshot 2024-03-22 at 2 11 48 PM" src="https://github.com/openwebwork/webwork2/assets/4732672/7eacd473-9384-49bf-b933-a2636674cdc9">

@drgrice1 , does anything occur to you about this? I am wondering if it makes sense for something in the feedback js and mqeditor.js could send iframeResizer a signal to recompute. Or if not, what could be done about this. I don't find posts about iframeResizer failing to respond to popovers.

**Comment by drgrice1** [comment]
Note that the MathQuill toolbar is scroll-able.  So what you show in your screenshot is what is expected there.  The user can scroll the toolbar to access all buttons.  The toolbar scrolls with the mouse wheel, with arrow buttons when the toolbar has focus, and it scrolls as keyboard focus moves through the buttons.

What I did for when problems are loaded like this in webwork2 so that all of the feedback popover could be viewed is I made it so that the iframe is vertically scroll-able.

**Comment by Alex-Jordan** [comment]
I found an `iframeResizer` option `heightCalculationMethod: 'lowestElement'` that pays a price in performance (not noticeable at this scale I think) and got the iframe to resize for the feedback popover as well as the mqeditor. The mqeditor still had scrollbars in both dimensions, but then I added `.quill-toolbar { box-sizing: content-box; }` and they went away too. Now it all looks like:

<img width="640" alt="Screenshot 2024-03-22 at 10 20 04 PM" src="https://github.com/openwebwork/pg/assets/4732672/28a558a8-414f-45ef-972a-236beedb111e">

<img width="611" alt="Screenshot 2024-03-22 at 10 20 14 PM" src="https://github.com/openwebwork/pg/assets/4732672/3c8b1f9f-0cd6-4f06-b369-4389c1b1b06e">

That's all changes on the PTX side, but is it worth cosidering adding that CSS change to `pg`?

**Comment by drgrice1** [comment]
Maybe.  Although, I actually don't want the iframe to resize for the MathQuill toolbar.  It is designed to work in short windows, and I think that is a massive resize that looks ugly with all of the empty space that it creates.  There is nothing wrong with scrolling for tall things like that.

**Comment by Alex-Jordan** [comment]
The solution here for the popover also resizes for the mqeditor, so I'm living with that excess space as a side effect. I would not object to the vertical scrolling if there were some other way to resize the iframe for the popover.

Note in the first picture from the opening post here that the mqeditor also has horizontal scrolling, which seems unnecessary. The CSS PTX is now applying also makes that go away.

**Comment by drgrice1** [comment]
It would be ideal if the iframe would resize for the popover, but not for the MathQuill toolbar.  It may be possible to do that with some care.  Perhaps this could be attained with `heightCalculationMethod: 'taggedElement'` and some effort tagging elements appropriately.  It certainly could be done with more effort using our own custom sizing method.

I am not sure where the horizontal scrolling is coming from in your screenshot.  That doesn't happen in webwork2.  The `mqeditor.css` file only sets `overflow-y: auto`, and does not set `overflow-x` for the `.quill-toolbar`.  So that may be some other style on your page.

**Comment by drgrice1** [comment]
I did a little testing, and if you add `data-iframe-height="1"` to the main `div` in the `templates/RPCRenderFormat/default.html.ep` file, and add `data-iframe-height="1"` to the popover tip via javascript in `htdocs/js/Feedback/feedback.js`, and add the `heightCalculationMethod: 'taggedElement'` option to the `iFrameResize` call in `htdocs/js/RenderProblem/renderproblem.js`, (probably also change `html{overflow-y:auto}` back to `html{overflow-y:hidden}` in that last file on line 31) then things work rather nicely.  The iframe resizes when a feedback popover opens, but not for the MathQuill toolbar, and all other content fits well.

**Comment by Alex-Jordan** [comment]
I tried these changes:

```
diff --git a/htdocs/js/Feedback/feedback.js b/htdocs/js/Feedback/feedback.js
index 1b6bccc..ca10438 100644
--- a/htdocs/js/Feedback/feedback.js
+++ b/htdocs/js/Feedback/feedback.js
@@ -44,6 +44,8 @@
                                ?.querySelector('.popover-header')
                                ?.addEventListener('click', () => feedbackPopover.hide());
 
+                       if (feedbackPopover.tip) feedbackPopover.tip.dataset.iframeHeight = "1";
+
                        const revealCorrectBtn = feedbackPopover.tip?.querySelector('.reveal-correct-btn');
                        if (revealCorrectBtn && feedbackPopover.correctRevealed) {
                                revealCorrectBtn.nextElementSibling?.classList.remove('d-none');

```

```
diff --git a/htdocs/js/RenderProblem/renderproblem.js b/htdocs/js/RenderProblem/renderproblem.js
index 9bdd58b..bc35c22 100644
--- a/htdocs/js/RenderProblem/renderproblem.js
+++ b/htdocs/js/RenderProblem/renderproblem.js
@@ -28,7 +28,7 @@
                                send_pg_flags: 1,
                                extra_header_text:
                                        '<style>' +
-                                       'html{overflow-y:auto;}body{padding:1px;background:#f5f5f5;}.container-fluid{padding:0px;}' +
+                                       'html{overflow-y:hidden;}body{padding:1px;background:#f5f5f5;}.container-fluid{padding:0px;}' +
                                        '</style>',
                                ...renderOptions
                        };
@@ -76,7 +76,7 @@
                                                        container.innerHTML = data.pg_flags.comment;
                                                        iframe.after(container);
                                                }
-                                               iFrameResize({ checkOrigin: false, warningTimeout: 20000, scrolling: 'omit' }, iframe);
+                                               iFrameResize({ checkOrigin: false, warningTimeout: 20000, scrolling: 'omit', heightCalculationMethod: 'taggedElement' }, iframe);
                                                iframe.addEventListener('load', () => resolve());
                                        }
                                        iframe.srcdoc = data.html;
```

```
diff --git a/templates/RPCRenderFormats/default.html.ep b/templates/RPCRenderFormats/default.html.ep
index 4007388..0e19edf 100644
--- a/templates/RPCRenderFormats/default.html.ep
+++ b/templates/RPCRenderFormats/default.html.ep
@@ -34,7 +34,7 @@
        %== $extra_header_text
 </head>
 <body>
-       <div class="container-fluid">
+       <div class="container-fluid" data-iframe-height="1">
                <div class="row g-0">
                        <div class="col-12 problem">
                                %== $resultSummary
```

along with removing the changes I previously mentioned to `pretext-webwork.js`. What ended up happening is the iframe came out with 0 height, and nothing was visible.

Maybe next week at a meeting we can compare notes?

**Comment by drgrice1** [comment]
Those are pretty much the changes that I made.  In webwork2 on the set detail page, library browser, problem grader page, and stats page, (all of the places where this is used) this works rather well.  Note that the PG problem editor does not use the IFrameResizer, so this does not apply there.

What version of IFrameResizer are you using for PreTeXt?  Maybe it is an older version that doesn't have the `taggedElement` option?  It looks like it was added in version 3.2.0.  Webwork is using 4.3.7 (although that will probably be updated to 4.3.9 before the release -- the latest version).

**Comment by drgrice1** [comment]
I just noticed that if you don't add the `data-iframe-height` attribute to the containing `div` in `default.html.ep`, then the iframe resizes to a height of zero.  Are you actually using that containing `div` in PreTeXt, or are you replacing that?  Are you not using the `raw` format anymore?  If you are using the `raw` format, then you would not get what is in the `default.html.ep` file, and would need to add the data attribute to your own container.

**Comment by Alex-Jordan** [comment]
I have this pretty ironed out now, and for example it looks like this with the popover extending the iframe height, but not the mqeditor:

<img width="651" alt="Screenshot 2024-03-25 at 12 11 51 AM" src="https://github.com/openwebwork/pg/assets/4732672/dbc7a1c4-790a-429d-bfa6-a1b67ad538da">

I needed to add two pieces of styling to the toolbar. Right now both are added at a certain place in the PTX repository (well, for now on my local branch). But I wanted to ask if you think either is appropriate for somwhere in `pg`. They are:

```
.quill-toolbar {
    overflow-x: hidden;
    scrollbar-width: thin;
}
```

Without the `overflow-x: hidden`, I get a horizontal scrollbar at the bottom (at least, in FF). It seems that the presence of the vertical scrollbar covers content and then FF wants to make a horizontal scrollbar too to account for that.

The `scrollbar-width: thin` makes the scrollbar thin, of course. According to caniuse, it is not respected in Safari.

After your thoughts on those two pieces of CSS, I will open small PRs to `webwork2` and `pg` that relate to this.

**Comment by drgrice1** [comment]
Those should be fine.  The `overflow-x: hidden` style isn't needed for webwork2, but doesn't interfere (or change anything) as far as I can tell.  However, for Firefox on linux an overlay scrollbar is used, and I notice that with the `scrollbar-width: thin` that overlay scrollbar is quite thin.  Screenshots are below (first is with `thin` and second without).  This is not critical since I doubt students really use linux.

![scrollbar-thin](https://github.com/openwebwork/pg/assets/47527406/53a20aec-ac99-4f2c-bab7-d73a15d2e645)
![scrollbar-regular](https://github.com/openwebwork/pg/assets/47527406/91059ce4-c687-4162-b020-8f435c559dbe)

**Comment by Alex-Jordan** [comment]
I notice in your screenshot that the scrollbar is transparent. You can see on my screenshot that it isn't. That's Firefox on Mac.

**Comment by drgrice1** [comment]
Yeah, that is what I meant by an overlay scrollbar.  It is transparent and overlays the region it scrolls.  I guess transparency is not really part of "overlay". I'm any case that is how it is on Linux (at least Ubuntu with Gnome).  When it is thin it becomes hard to see.

**Comment by Alex-Jordan** [comment]
I should have checked this earlier, but I checked the Library Browser (on develop with the small changes already mentioned in this thread) and I see the same that I see with PTX:

<img width="1317" alt="Screenshot 2024-03-25 at 12 31 08 PM" src="https://github.com/openwebwork/pg/assets/4732672/fd2995eb-8039-4a82-a910-f16a3cef74c9">

That is, I see the horizontal scrollbar, and neither scrollbar is transparent. So I guess we see different things in Firefox becasue of Mac vs Linux. The horizontal scrollbar is pretty annoying, and the `overflow-x: hidden` makes it go away. So I feel like I should add that. But I could go either way on the thin width. It's not thin in this screenshot here, but it is in my previous screenshot, so you can compare.

**Comment by drgrice1** [comment]
I don't care on the thin scrollbar.  For Firefox or Chrome on Linux there is no horizontal scrollbar.  But if it is showing up on other platforms, the the `overflow-x: hidden` is needed.

**Comment by Alex-Jordan** [comment]
This was addressed with #1045.
